### PR TITLE
avx2 for hmmer

### PR DESCRIPTION
**Pull request by mr-c**

### CODE REVIEW

**Comment by mr-c** [comment]
Hello @nemequ, any advice for the pgcc test failures?

**Comment by nemequ** [comment]
It looks like it's implemented in PGI with `__builtin_shufflevector`, but they use `_mm256_undefined_si256()` for the second argument instead of using `a` twice like you do here:

```c
#define _mm256_shuffle_epi32(a, imm) __extension__ ({ \
  (__m256i)__builtin_shufflevector((__v8si)(__m256i)(a), \
                                   (__v8si)_mm256_undefined_si256(), \
                                   0 + (((imm) >> 0) & 0x3), \
                                   0 + (((imm) >> 2) & 0x3), \
                                   0 + (((imm) >> 4) & 0x3), \
                                   0 + (((imm) >> 6) & 0x3), \
                                   4 + (((imm) >> 0) & 0x3), \
                                   4 + (((imm) >> 2) & 0x3), \
                                   4 + (((imm) >> 4) & 0x3), \
                                   4 + (((imm) >> 6) & 0x3)); })
```

As for `mm256_srli_si256`, their implementation is interesting but shouldn't be problematic:

```c
#define _mm256_srli_si256(a, imm) __extension__ ({ \
  (__m256i)__builtin_shufflevector(                                           \
        (__v32qi)(__m256i)(a),                                               \
        (__v32qi)_mm256_setzero_si256(),                                     \
        ((char)(imm)&0xF0) ? 32 : (char)(imm) + ((char)(imm)>0xF ? 16 : 0),  \
        ((char)(imm)&0xF0) ? 33 : (char)(imm) + ((char)(imm)>0xE ? 17 : 1),  \
        ((char)(imm)&0xF0) ? 34 : (char)(imm) + ((char)(imm)>0xD ? 18 : 2),  \
        ((char)(imm)&0xF0) ? 35 : (char)(imm) + ((char)(imm)>0xC ? 19 : 3),  \
        ((char)(imm)&0xF0) ? 36 : (char)(imm) + ((char)(imm)>0xB ? 20 : 4),  \
        ((char)(imm)&0xF0) ? 37 : (char)(imm) + ((char)(imm)>0xA ? 21 : 5),  \
        ((char)(imm)&0xF0) ? 38 : (char)(imm) + ((char)(imm)>0x9 ? 22 : 6),  \
        ((char)(imm)&0xF0) ? 39 : (char)(imm) + ((char)(imm)>0x8 ? 23 : 7),  \
        ((char)(imm)&0xF0) ? 40 : (char)(imm) + ((char)(imm)>0x7 ? 24 : 8),  \
        ((char)(imm)&0xF0) ? 41 : (char)(imm) + ((char)(imm)>0x6 ? 25 : 9),  \
        ((char)(imm)&0xF0) ? 42 : (char)(imm) + ((char)(imm)>0x5 ? 26 : 10), \
        ((char)(imm)&0xF0) ? 43 : (char)(imm) + ((char)(imm)>0x4 ? 27 : 11), \
        ((char)(imm)&0xF0) ? 44 : (char)(imm) + ((char)(imm)>0x3 ? 28 : 12), \
        ((char)(imm)&0xF0) ? 45 : (char)(imm) + ((char)(imm)>0x2 ? 29 : 13), \
        ((char)(imm)&0xF0) ? 46 : (char)(imm) + ((char)(imm)>0x1 ? 30 : 14), \
        ((char)(imm)&0xF0) ? 47 : (char)(imm) + ((char)(imm)>0x0 ? 31 : 15), \
        ((char)(imm)&0xF0) ? 48 : (char)(imm) + ((char)(imm)>0xF ? 32 : 16), \
        ((char)(imm)&0xF0) ? 49 : (char)(imm) + ((char)(imm)>0xE ? 33 : 17), \
        ((char)(imm)&0xF0) ? 50 : (char)(imm) + ((char)(imm)>0xD ? 34 : 18), \
        ((char)(imm)&0xF0) ? 51 : (char)(imm) + ((char)(imm)>0xC ? 35 : 19), \
        ((char)(imm)&0xF0) ? 52 : (char)(imm) + ((char)(imm)>0xB ? 36 : 20), \
        ((char)(imm)&0xF0) ? 53 : (char)(imm) + ((char)(imm)>0xA ? 37 : 21), \
        ((char)(imm)&0xF0) ? 54 : (char)(imm) + ((char)(imm)>0x9 ? 38 : 22), \
        ((char)(imm)&0xF0) ? 55 : (char)(imm) + ((char)(imm)>0x8 ? 39 : 23), \
        ((char)(imm)&0xF0) ? 56 : (char)(imm) + ((char)(imm)>0x7 ? 40 : 24), \
        ((char)(imm)&0xF0) ? 57 : (char)(imm) + ((char)(imm)>0x6 ? 41 : 25), \
        ((char)(imm)&0xF0) ? 58 : (char)(imm) + ((char)(imm)>0x5 ? 42 : 26), \
        ((char)(imm)&0xF0) ? 59 : (char)(imm) + ((char)(imm)>0x4 ? 43 : 27), \
        ((char)(imm)&0xF0) ? 60 : (char)(imm) + ((char)(imm)>0x3 ? 44 : 28), \
        ((char)(imm)&0xF0) ? 61 : (char)(imm) + ((char)(imm)>0x2 ? 45 : 29), \
        ((char)(imm)&0xF0) ? 62 : (char)(imm) + ((char)(imm)>0x1 ? 46 : 30), \
        ((char)(imm)&0xF0) ? 63 : (char)(imm) + ((char)(imm)>0x0 ? 47 : 31)); })
```

Interestingly, clang doesn't use `__builtin_shufflevector` for either.

So it should be possible to test both implementations using clang to see if they're correct.

(Note that the license on that header is a BSD, so pasting the code here should be fine).

**Comment by mr-c** [comment]
Thank you for the explanation, @nemequ but I'm still a bit lost :-P

Are you saying that I should just copy and paste these inside a `#if defined __PGI` block?
Where do I get the __m256i value to pass to the shufflevector? There is no ".n" in the `simde__m256i` struct when we lack AVX2.

Or did I misunderstand your suggestion in a very different way?

File: .travis.yml [hunk 1]
```diff
@@ -20,6 +20,15 @@ matrix:
           - libc6-dev-i386-cross
           - cmake
           - cmake-data
+    - env: C_COMPILER=i686-linux-gnu-gcc-8 CFLAGS=-march=i686 -DSIMDE_NO_SHUFFLE_VECTOR -m-no-avx2 -m-no-avx -m-no-sse4.2 -m-no-sse4.1 -m-no-ssse3 -m-no-sse3 -m-no-sse2 -m-no-sse
```
Comment (junaruga) [inline .travis.yml]: You can use `- env:` with hash value like this. And `CFLAGS` might need the single quote `'` or double quote `*`.

```
- env:
  - C_COMPILER=i686-linux-gnu-gcc-8
  - CFLAGS='-march=i686 -DSIMDE_NO_SHUFFLE_VECTOR -m-no-avx2 -m-no-avx -m-no-sse4.2 -m-no-sse4.1 -m-no-ssse3 -m-no-sse3 -m-no-sse2 -m-no-sse'
```

Comment (mr-c) [reply .travis.yml]: @junaruga Good catch, thanks!

**Comment by nemequ** [comment]
I was just trying to say that we could verify their implementation using clang to see if the problem was with it or the compiler.  It seems to work on clang, but it actually crashed PGI here so there is definitely something wrong with their internals.

We can't use `__builtin_shufflevector` for PGI because they don't fully support GCC vector extensions; you can't access the elements with square brackets.  Since we depend on that pretty heavily in SIMDe (if vector extensions are available we use them *instead of* an array, not in addition to one) it would be quite a bit of work to refactor and I don't think it's worth it.  If we had those arrays that's what you would pass to `__builtin_shufflevector`, not the `__m256i`.

You did the right thing by just blacklisting PGI.  I've had to do that a bunch of times, and I usually report the bugs I find in their forum, but they pretty much never get fixed.

**Pull request merged**